common/reviews/api/terminal.api.md

```diff
@@ -57,12 +159,47 @@ export interface IStdioSummarizerOptions {
     trailingLines?: number;
 }
 
+// @beta (undocumented)
+export interface IStringBufferOutputOptions {
+    normalizeSpecialCharacters: boolean;
+}
+
+// @beta (undocumented)
+export interface ITerminal {
+    registerProvider(provider: ITerminalProvider): void;
+    unregisterProvider(provider: ITerminalProvider): void;
+    write(...messageParts: WriteParameters): void;
+    writeDebug(...messageParts: WriteParameters): void;
+    writeDebugLine(...messageParts: WriteParameters): void;
+    writeError(...messageParts: WriteParameters): void;
+    writeErrorLine(...messageParts: WriteParameters): void;
+    writeLine(...messageParts: WriteParameters): void;
+    writeVerbose(...messageParts: WriteParameters): void;
+    writeVerboseLine(...messageParts: WriteParameters): void;
```

Comment:
Code that calls `writeDebug` or `writeVerbose` may involve nontrivial performance overhead, not just for generating lots of strings, but also possibly for querying whatever information needs to be formatted in those strings.

Is it inherently inefficient that the caller cannot do something like this?
```ts
if (terminal.hasVerbose) {
  terminal.writeVerbose(generateVerboseReport());
}
if (terminal.hasDebug) {
  const report: string = doSuperExpensiveThing();
  terminal.writeDebug(report);
}
```

I understand that these states are not always globally controlled by `--verbose` or `--debug`. It may be the case that multiple providers are listening, and one of them wants verbose but the other does not.  But in that case the `hasVerbose` property would return `true` if we know that ANY of the providers would consume verbose inputs.

Reply:
This is a bit out of scope for your PR, but we are doing a MAJOR version bump.... 😁

Reply:
That would be a pretty significant change to the currently-simple `ITerminalProvider` interface. We should consider doing that in a follow-up before major bumping `@rushstack/terminal`.